components/brave_wallet/browser/zcash/rust/cxx_orchard_shard_tree_delegate.cc

```diff
@@ -0,0 +1,298 @@
+}
+
+CxxOrchardShardAddress ToRust(const ::brave_wallet::OrchardShardAddress& addr) {
+  return CxxOrchardShardAddress{addr.level, addr.index};
+}
+
+CxxOrchardShardTreeCap ToRust(
+    ::brave_wallet::OrchardShardTreeCap& shard_store_cap) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(shard_store_cap.size());
+  base::ranges::copy(shard_store_cap, std::back_inserter(data));
+  return CxxOrchardShardTreeCap{std::move(data)};
+}
+
+::brave_wallet::OrchardShardTreeCap FromRust(
+    const CxxOrchardShardTreeCap& cap) {
+  ::brave_wallet::OrchardShardTreeCap shard_store_cap;
+  shard_store_cap.reserve(cap.data.size());
+  base::ranges::copy(cap.data, std::back_inserter(shard_store_cap));
+  return shard_store_cap;
+}
+
+::brave_wallet::OrchardShard FromRust(const CxxOrchardShard& tree) {
+  std::optional<OrchardShardRootHash> shard_root_hash;
+  if (!tree.hash.empty()) {
+    CHECK_EQ(kOrchardShardTreeHashSize, tree.hash.size());
+    OrchardShardRootHash hash_value;
+    base::ranges::copy(tree.hash, hash_value.begin());
+    shard_root_hash = hash_value;
+  }
+
+  std::vector<uint8_t> data;
+  data.reserve(tree.data.size());
+  base::ranges::copy(tree.data, std::back_inserter(data));
+
+  return ::brave_wallet::OrchardShard(FromRust(tree.address), shard_root_hash,
+                                      std::move(data));
+}
+
+CxxOrchardShard ToRust(const ::brave_wallet::OrchardShard& tree) {
+  ::rust::Vec<uint8_t> data;
+  data.reserve(tree.shard_data.size());
+  base::ranges::copy(tree.shard_data, std::back_inserter(data));
+
+  ::rust::Vec<uint8_t> hash;
+  if (tree.root_hash) {
+    base::ranges::copy(tree.root_hash.value(), std::back_inserter(hash));
+  }
+  return CxxOrchardShard{ToRust(tree.address), std::move(hash),
+                         std::move(data)};
+}
+
+CxxOrchardCheckpoint ToRust(
+    const ::brave_wallet::OrchardCheckpoint& checkpoint) {
+  ::rust::Vec<uint32_t> marks_removed;
+  base::ranges::copy(checkpoint.marks_removed,
+                     std::back_inserter(marks_removed));
+  return CxxOrchardCheckpoint{!checkpoint.tree_state_position.has_value(),
+                              checkpoint.tree_state_position.value_or(0),
+                              marks_removed};
+}
+
+CxxOrchardCheckpointBundle ToRust(
+    const ::brave_wallet::OrchardCheckpointBundle& checkpoint_bundle) {
+  return CxxOrchardCheckpointBundle(checkpoint_bundle.checkpoint_id,
+                                    ToRust(checkpoint_bundle.checkpoint));
+}
+
+::brave_wallet::OrchardCheckpoint FromRust(
+    const CxxOrchardCheckpoint& checkpoint) {
+  CheckpointTreeState checkpoint_tree_state = std::nullopt;
+  if (!checkpoint.empty) {
+    checkpoint_tree_state = checkpoint.position;
+  }
+  return ::brave_wallet::OrchardCheckpoint{
+      checkpoint_tree_state,
+      std::vector<uint32_t>(checkpoint.mark_removed.begin(),
+                            checkpoint.mark_removed.end())};
```

Comment:
Could we avoid this copy by having `checkpoint` being passed by value?

Reply:
Here is conversion between different types. Between OrchardCheckpoint and CxxOrchardCheckpoint

Reply:
Ack.